From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 28 Jan 2023 14:10:43 -0500
Subject: [PATCH] Dont close Slime Plugin Classloader


diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index fe21d31b463317eb90d58cbca5f098958ca93938..8ae0addea5a00938eb04169117ccfb6cecae2d75 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -256,7 +256,9 @@ public final class PluginClassLoader extends URLClassLoader implements io.paperm
             super.close();
         } finally {
             this.closed = true; // Pufferfish
+            if (!this.plugin.getName().equals("SlimeWorldManager")) { // ASWM - Don't close
             jar.close();
+            } // ASWM - Don't close
         }
     }
 
