From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomTom <93038247+AverageGithub@users.noreply.github.com>
Date: Wed, 29 Mar 2023 18:38:48 +0200
Subject: [PATCH] Disable out-of order chat kicking


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 2d8488b9aa088b6d5f0c7e557b8ad0b29bc4cd88..e6cb20f15761fc240780435d4e276a58adb86d02 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2346,15 +2346,16 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     }
 
     private boolean updateChatOrder(Instant timestamp) {
-        Instant instant1;
-
-        do {
-            instant1 = (Instant) this.lastChatTimeStamp.get();
-            if (timestamp.isBefore(instant1)) {
-                return false;
-            }
-        } while (!this.lastChatTimeStamp.compareAndSet(instant1, timestamp));
-
+// RivalsBW start - Disable out-of-order chat kick
+//        Instant instant1;
+//
+//        do {
+//            instant1 = (Instant) this.lastChatTimeStamp.get();
+//            if (timestamp.isBefore(instant1)) {
+//                return false;
+//            }
+//        } while (!this.lastChatTimeStamp.compareAndSet(instant1, timestamp));
+// RivalsBW end
         return true;
     }
 
