From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomTom <93038247+AverageGithub@users.noreply.github.com>
Date: Wed, 29 Mar 2023 19:09:19 +0200
Subject: [PATCH] Another performance boost!


diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index fbe209a66c77c47935ad026dd3e45e682af91fd8..26a5ca999e8bc0f2ba13e2d6bb74f53437d0a590 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1617,11 +1617,11 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
                 double d2 = d0 * d0;
                 boolean flag = d1 <= d2 && this.entity.broadcastToPlayer(player);
 
-                // CraftBukkit start - respect vanish API
-                if (!player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) {
-                    flag = false;
-                }
-                // CraftBukkit end
+                // CraftBukkit start - respect vanish API // RivalsBW start - Disable canSee
+//                if (!player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) {
+//                    flag = false;
+//                }
+                // CraftBukkit end // RivalsBW end - Disable canSee
                 if (flag) {
                     if (this.seenBy.add(player.connection)) {
                         this.serverEntity.addPairing(player);
