From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Github Actions <no-reply@github.com>
Date: Thu, 30 Mar 2023 15:42:37 +0200
Subject: [PATCH] Disable unneeded entity actions


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 4705d7066207250c03a5f98eef61554c901f2e35..579fc54f97a84cfccbca25444e0163d67665610f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -810,9 +810,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
         this.wasInPowderSnow = this.isInPowderSnow;
         this.isInPowderSnow = false;
-        this.updateInWaterStateAndDoFluidPushing();
-        this.updateFluidOnEyes();
-        this.updateSwimming();
+//        this.updateInWaterStateAndDoFluidPushing();  // RivalsBW start - Disable unneeded entity actions
+//        this.updateFluidOnEyes();
+//        this.updateSwimming(); // RivalsBW end - Disable unneeded entity actions
         if (this.level.isClientSide) {
             this.clearFire();
         } else if (this.remainingFireTicks > 0) {
