From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Github Actions <no-reply@github.com>
Date: Fri, 31 Mar 2023 16:47:25 +0200
Subject: [PATCH] mhm


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 4c9784629c89966c0b51d809bbfea6ef8d8c7c97..b7e014ed0c34f38b36d9bcc87f3f40152be237fe 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -315,7 +315,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public volatile Thread shutdownThread; // Paper
     public volatile boolean abnormalExit = false; // Paper
     public boolean isIteratingOverLevels = false; // Paper
-    public hu.rivalsnetwork.AsyncUtils asyncUtils = new hu.rivalsnetwork.AsyncUtils(); // RivalsBW
+    public hu.rivalsnetwork.AsyncUtils asyncUtils = new hu.rivalsnetwork.AsyncUtils("PlayerUpdateThread"); // RivalsBW
 
     public static <S extends MinecraftServer> S spin(Function<Thread, S> serverFactory) {
         AtomicReference<S> atomicreference = new AtomicReference();
